pkg/cmd/roachtest/tests/mixed_version_tenant_span_stats.go

```diff
@@ -150,34 +150,40 @@ func registerTenantSpanStatsMixedVersion(r registry.Registry) {
 					}
 				} else {
 					// All nodes are on one version, but we're in mixed state (i.e. cluster version is on a different version)
-					var issueNodeID int
-					var dialNodeID int
-					// All nodes on current version
-					if len(h.Context().ToVersionNodes) == 4 {
-						issueNodeID = h.Context().ToVersionNodes[0]
-						dialNodeID = h.Context().ToVersionNodes[1]
-					} else {
-						// All nodes on previous version
-						issueNodeID = h.Context().FromVersionNodes[0]
-						dialNodeID = h.Context().FromVersionNodes[1]
+					issueNodeID := c.All()[0]
+					dialNodeID := c.All()[1]
+					var clusterVersion string
+
+					if err = h.QueryRow(rng, `SHOW CLUSTER SETTING version`).Scan(&clusterVersion); err != nil {
+						return err
 					}
-					// Dial a node for span stats.
-					l.Printf("Dial a node for span stats (different cluster version).")
-					res, err = fetchSpanStatsFromNode(ctx, l, c, c.Node(issueNodeID), newReqBody(dialNodeID, startKey, endKey))
+					cv, err := version.Parse(clusterVersion)
 					if err != nil {
 						return err
 					}
-					// Expect an error in the stdout - mixed version error.
-					// Ensure the result can be marshalled into a valid error response.
-					err = json.Unmarshal([]byte(res.Stdout), &errOutput)
+
+					// Dial a node for span stats.
+					l.Printf("Dial a node for span stats (different cluster version).")
+					res, err = fetchSpanStatsFromNode(ctx, l, c, c.Node(issueNodeID), newReqBody(dialNodeID, startKey, endKey))
 					if err != nil {
 						return err
 					}
-					// Ensure we get the expected error.
-					mixedClusterVersionErr := assertExpectedError(errOutput.Message, mixedVersionReqError)
-					expectedUnknown := assertExpectedError(errOutput.Message, unknownFieldError)
-					if !mixedClusterVersionErr && !expectedUnknown {
-						return errors.Newf("expected '%s' or '%s' in error message, got: '%v'", mixedVersionReqError, unknownFieldError, errOutput.Error)
+					// An error is expected if:
+					// - the cluster version is <23.1.0
+					// - or cluster version >=23.1 and node versions <23.1.0.
+					if !cv.AtLeast(v231) || len(h.Context().FromVersionNodes) == 4 {
```

Comment:
A few comments.

* `cluster version >=23.1` is misleading as, in this test, the cluster version will be at most V23_1 by definition.
* `cluster version >=23.1 and node versions <23.1.0` -- not sure what you meant here, but cluster version being in 23.1 and nodes being <23.1 is not possible. Upgrades will only happen once all nodes are running the 23.1 binary and once the upgrade begins, older binaries are no longer able to join.
* There's a TOCTOU issue with regards to the cluster version: if this function is called _while the cluster is finalizing_, there's no guarantees the value you're using here still reflects the actual cluster version. I believe that if you run this test using a seed that will lead to this function being called during finalization, and sleep for 3mins on line 164, this test will fail with high probability (i.e., there's a timing dependency).

I think the issue with the test is that it may flake if this hook is called during finalization. Does it make sense to even perform the error check in this case? I'm having a hard time thinking of a meaningful assertion we can write in that environment.

If you decide to skip, you can add the following to the top of this function:

```go
if h.Context().Finalizing {
    return nil
}
```

Reply:
In this case, can the versioning logic simply be gated to the cluster version? Do we need to check the node versions at all?

Reply:
Could you elaborate? What "versioning logic" are you referring to?

> Do we need to check the node versions at all?

In what context?

*****

To make myself clearer (in case it wasn't), my suggestion is to replace the entire diff in this PR with the above check for `Finalizing` in the first line of the `fetch span stats - mixed` function. As I said, it will be quite tricky, if not impossible, to write a non-flaky assertion while the upgrade is finalizing.

Reply:
Sorry, was in a couple long pairing sessions.

> Could you elaborate? What "versioning logic" are you referring to?
Currently, in the test, there is a check to see if the nodes are on different binaries. If so, we check that we receive the expected error message. If the cluster version is only at 23.1.0 when all nodes are are at least at 23.1.0, then I suspect we don't need this check at all? We can just use a cluster version check.

> To make myself clearer (in case it wasn't), my suggestion is to replace the entire diff in this PR with the above check for Finalizing in the first line of the fetch span stats - mixed function. As I said, it will be quite tricky, if not impossible, to write a non-flaky assertion while the upgrade is finalizing.

For my own knowledge, what does it _mean_ for an upgrade to be finalizing, could it be described as when the cluster version is upgraded once all nodes have upgraded?

Reply:
> why the diff I've written would cause flakes.

It causes flakes in the following scenario:

1. the `fetch span stats - mixed` function is called while the upgrade is finalizing (all nodes running a `release-23.1` binary, migrations are running in the background, cluster version is advancing from `V22_2` all the way to `V23_1` _while_ your function is running)
2. You read the current cluster version in "fetch span stats". Let's say that that returns `22.2-86` (i.e., by the time you read the cluster setting, quite a few migrations have completed).
3. Since the cluster version is <23.1, you deem that you should get an error if you try to use your endpoint.
4. In the meantime, the remaining migrations finish running, and the cluster version reaches `V23_1`
5. You call your endpoint. You expect an error, but no error is returned (behavior is correct, assertion is wrong).

Reply:
With that in mind, does it make sense to strip the entire `else` branch?

I.e.
```
} else {
					// All nodes are on one version, but we're in mixed state (i.e. cluster version is on a different version)
					issueNodeID := c.All()[0]
					dialNodeID := c.All()[1]

					// Dial a node for span stats.
					l.Printf("Dial a node for span stats (different cluster version).")
					res, err = fetchSpanStatsFromNode(ctx, l, c, c.Node(issueNodeID), newReqBody(dialNodeID, startKey, endKey))
					if err != nil {
						return err
					}

					// Expect an error in the stdout - mixed version error.
					// Ensure the result can be marshalled into a valid error response.
					err = json.Unmarshal([]byte(res.Stdout), &errOutput)
					if err != nil {
						return err
					}
					// Ensure we get the expected error.
					mixedClusterVersionErr := assertExpectedError(errOutput.Message, mixedVersionReqError)
					expectedUnknown := assertExpectedError(errOutput.Message, unknownFieldError)
					if !mixedClusterVersionErr && !expectedUnknown {
						return errors.Newf("expected '%s' or '%s' in error message, got: '%v'", mixedVersionReqError, unknownFieldError, errOutput.Error)
					}
				}
```
Is there a state where all nodes are on one version, the cluster is on another, but we aren't in a finalizing state?

Reply:
Agreed -- if we skip on `Finalizing`, we can remove the entire `else` branch -- we always expect an error in that case, regardless of which node we connect to.

Reply:
Sweet, made the change and updated the commit/PR desc to reflect the new diff.